scalafix-rules/src/main/scala-3/scalafix/internal/rule/ExplicitResultTypes.scala

```diff
@@ -0,0 +1,84 @@
+package scalafix.internal.rule
+
+import scala.meta.*
+
+import buildinfo.RulesBuildInfo
+import dotty.tools.pc.ScalaPresentationCompiler
+import metaconfig.Configured
+import scalafix.internal.pc.PcExplicitResultTypes
+import scalafix.patch.Patch
+import scalafix.v1.*
+
+final class ExplicitResultTypes(
+    val config: ExplicitResultTypesConfig,
+    fallback: Option[PcExplicitResultTypes]
+) extends SemanticRule("ExplicitResultTypes")
+    with ExplicitResultTypesBase[Scala3Printer] {
+
+  def this() = this(ExplicitResultTypesConfig.default, None)
+
+  val compilerScalaVersion: String = RulesBuildInfo.scalaVersion
+
+  private def toBinaryVersion(v: String) = v.split('.').take(2).mkString(".")
+
+  override def description: String =
+    "Inserts type annotations for inferred public members."
+
+  override def isRewrite: Boolean = true
+
+  override def afterComplete(): Unit = {
+    shutdownCompiler()
+  }
+
+  private def shutdownCompiler(): Unit = {
+    fallback.foreach(_.shutdownCompiler())
+  }
+
+  override def withConfiguration(config: Configuration): Configured[Rule] = {
+    config.conf // Support deprecated explicitReturnTypes config
+      .getOrElse("explicitReturnTypes", "ExplicitResultTypes")(
+        ExplicitResultTypesConfig.default
+      )
+      .map(c =>
+        new ExplicitResultTypes(
+          c,
+          Option {
+            if (
+              toBinaryVersion(config.scalaVersion) == toBinaryVersion(
+                compilerScalaVersion
+              )
+            )
+              PcExplicitResultTypes
+                .static(config, new ScalaPresentationCompiler())
+            else
+              PcExplicitResultTypes.dynamic(config)
```

Comment:
I think we could condition that to a configuration flag that makes the user aware that it's sub-optimal. If that flag is not set and we come here, we can provide an actionnable error message.